### ISSUE TITLE
Bagging - train final models with all training dataset available? 

### ISSUE THREAD

**Issue by yitang** [andyh47/higgs#1]
Hi Andy, I am new to Kaggle competition and found your scripts really helpful to learn those crazy ideals.

I have been reading the `withBagging-xgboost.py` and it seems the models in the bag are fitted with only 90% of the training dataset. Those models then are directly applied on the test dataset. The 10% of training dataset are left out, and the validation dataset is used for measuring the bag model only.

This characteristic is different from what I learnt about single model, in which case, once the optimal parameters are found using CV, we re-build the model again with all the dataset available , (100% training dataset + validation dataset), and then use the resulting model to predict.

I wonder if I misunderstood your scripts, or we just use bag model differently when applying to test dataset?  

Yi

**Comment by andyh47** [comment]
I'm glad you've found the script helpful.   You are correct that rebuilding
with the whole training set is a good idea.  This was a difficult
competition and I was spending more time worrying about the difference
between cross validated results and leaderboard results and incorporating
new data that arrived ~48 hours before the end of the competition (
submitted to all by a contestant).  So things were not optimal.

In general in these competitions, one has to worry a lot about
overfitting.  A better approach would have been to split into a training
set and 2 validation sets.  validation set 1 to be used on
hyper-opt/xgboost cross-validation and the second on cross-validating the
results of the bagging.   Bagging proved to be extremely helpful in getting
around overfitting and getting a good final position.

If faced with the cross-validation problem again, I would probably use
random-split, say 20 times, instead of 5x or 10x cross-validation.

Regards

On Thu, Jan 7, 2016 at 2:44 AM, Yi Tang notifications@github.com wrote:

> Hi Andy, I am new to Kaggle competition and found your scripts really
> helpful to learn those crazy ideals
> 
> I have been reading the withBagging-xgboostpy and it seems the models in
> the bag are fitted with only 90% of the training dataset Those models then
> are directly applied on the test dataset The 10% of training dataset are
> left out, and the validation dataset is used for measuring the bag model
> only
> 
> This characteristic is different from what I learnt about single model, in
> which case, once the optimal parameters are found using CV, we re-build the
> model again with all the dataset available , (100% training dataset +
> validation dataset), and then use the resulting model to predict
> 
> I wonder if I misunderstood your scripts, or we just use bag model
> differently when applying to test dataset?
> 
> Yi
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/andyh47/higgs/issues/1.

## 

Andy Haines
650 245 2691
andrewhaines71@gmail.com

**Comment by yitang** [comment]
Thanks for your confirmation, it saves me few hours in googleing. 

> I'm glad you've found the script helpful. You are correct that
> rebuilding with the whole training set is a good idea. This was a
> difficult competition and I was spending more time worrying about the
> difference between cross validated results and leaderboard results and
> incorporating new data that arrived ~48 hours before the end of the
> competition ( submitted to all by a contestant). So things were not optimal.

I just confronted a massive over-fitting
problem after running various models for 50 hours. I did a scatterplot
of the CV results and leaderboard results, for most of the points, the
leaderboard results are better than CV results, which is expected if
the model are good and trained with more data. But for 5 models, the
CV results are about 0.96 and the leaderboard results are about 0.85,
which really shocks me. Obviously I need to investigate this problem
but I don't have a clue at this moment about what to check. Could you please 
advice few things? I am looking for a technique to check over-fitting without 
getting the leaderboard results. 

> In general in these competitions, one has to worry a lot about
> overfitting.  A better approach would have been to split into a training
> set and 2 validation sets.  validation set 1 to be used on
> hyper-opt/xgboost cross-validation and the second on cross-validating the
> results of the bagging.   Bagging proved to be extremely helpful in getting
> around overfitting and getting a good final position.

Thanks for sharing me this insights. I'd like to apply this technique
to the homesite competition I am in and see how does it performs.

> If faced with the cross-validation problem again, I would probably use
> random-split, say 20 times, instead of 5x or 10x cross-validation.

Most of the machine learning books recommends 5 or 10 folder CV. I
wonder why you'd prefer random split? I tried to use random-split in
tuning the hyper-parameters, it tells a completely different story to
k-folder CV; given the same parameters, the results of random split
has much more variations, and the average is far away from the average
of CV estimates. Is this the desired characteristics we are looking
for in dealing with over-fitting?

Regards,

Yi

**Comment by andyh47** [comment]
There are two questions here.  Estimating performance with out looking at
the leaderboard and overfitting.

Taking performance estimation first; usually I use 5 fold cross
validation.  This is fast and works well enough.  The expected quality of
these techniques is given by the Standard Error = sample standard deviation
of estimates/ sqrt(number of estimates).   If the estimates have a high
standard deviation then n=5 or 10 may not be large enough.  Going with
k-fold with k large say 100 means that 99% of the data is used for training
and 1% for test. This is not optimal.  Test is so small that sampling
effects can confuse results.  With shuffle split you can set the split to
80/20 like k=5 and then run 20 times or 40 or whatever.   This takes a lot
of time but may be necessary.   In the Higgs competition one contestant ran
5-fold 4 times with a shuffle before each one.  Another option.

Shuffle/split resamples from the data as you would doing a bootstrap.
K-fold does not.  My expectation is that the results might be a different
for small n but will converge for large n.

Re: overfitting.  The key to overfitting is vigilance.  Overfitting usually
arrises because you are looking at all the data during training in some
way.  When you do a cross validation, look at the results and then tweak
some parameters and re-run, you are effectively taking all the data into
account in the training.  When you do this 100 times during a competition
the effect can be large. Having a hold-out set (train,test, validation)
 which is used sparingly is really essential since there are so many model
iterations in the typical contest.

The reason that I say vigilance is important is that 'peeking' at the data
can happen in subtle ways. In natural language processing, the first step
is to convert text to a matrix using bag of words.  The temptation is to do
the conversion then use cross validation on the matrix.   However, the text
to matrix operation looks a ALL of the words so both the train and test
matrices effectively contain some information from all of the data.  The
right way to do this is put the conversation step inside the cross
validation loop.  The effect in this case is usually not large but it is
there.

In most competitions, having a hierarchical model, with some sort of
stacking is required to win. This presents another opportunity to peek when
optimizing the parameters for each level of the hierarchy.  If there is
enough data having a holdout set for each level is the way to go.  E.g.
(train,test, validation1, validation2) etc.

I use the word vigilance because information leaking into training can be
subtle. Each time you add a processing step ask your self if any
information from test is being used during the train processing.

On Mon, Jan 11, 2016 at 1:02 AM, Yi Tang notifications@github.com wrote:

> Thanks for your confirmation, it saves me few hours in googleing.
> 
> I'm glad you've found the script helpful. You are correct that
> rebuilding with the whole training set is a good idea. This was a
> difficult competition and I was spending more time worrying about the
> difference between cross validated results and leaderboard results and
> incorporating new data that arrived ~48 hours before the end of the
> competition ( submitted to all by a contestant). So things were not
> optimal.
> 
> I just confronted a massive over-fitting
> problem after running various models for 50 hours. I did a scatterplot
> of the CV results and leaderboard results, for most of the points, the
> leaderboard results are better than CV results, which is expected if
> the model are good and trained with more data. But for 5 models, the
> CV results are about 0.96 and the leaderboard results are about 0.85,
> which really shocks me. Obviously I need to investigate this problem
> but I don't have a clue at this moment about what to check. Could you
> please
> advice few things? I am looking for a technique to check over-fitting
> without
> getting the leaderboard results.
> 
> In general in these competitions, one has to worry a lot about
> overfitting. A better approach would have been to split into a training
> set and 2 validation sets. validation set 1 to be used on
> hyper-opt/xgboost cross-validation and the second on cross-validating the
> results of the bagging. Bagging proved to be extremely helpful in getting
> around overfitting and getting a good final position.
> 
> Thanks for sharing me this insights. I'd like to apply this technique
> to the homesite competition I am in and see how does it performs.
> 
> If faced with the cross-validation problem again, I would probably use
> random-split, say 20 times, instead of 5x or 10x cross-validation.
> 
> Most of the machine learning books recommends 5 or 10 folder CV. I
> wonder why you'd prefer random split? I tried to use random-split in
> tuning the hyper-parameters, it tells a completely different story to
> k-folder CV; given the same parameters, the results of random split
> has much more variations, and the average is far away from the average
> of CV estimates. Is this the desired characteristics we are looking
> for in dealing with over-fitting?
> 
> Regards,
> 
> Yi
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/andyh47/higgs/issues/1#issuecomment-170473449.

## 

Andy Haines
650 245 2691
andrewhaines71@gmail.com
